include/merlin/memory_pool.hpp

```diff
@@ -0,0 +1,600 @@
+ *
+ * When borrowing a buffer a streaming context can be specified. This context is
+ * relevant for allocation and during returns. It is assumed that the stream you
+ * provide as context will be the stream where you queue the workload. Not doing
+ * so may lead to undefined behavior.
+ *
+ * Buffer return:
+ *
+ * If no context is provided, we cannot make any assumptions regarding the usage
+ * one the device. So we sychronize the device first and then return the buffer
+ * to the stock. If a streaming context was provided, we queue an event and add
+ * the buffer to the `pending` pool. That means, the buffer has been
+ * reqlinquished by the CPU, but may still be used by the GPU. If no pending
+ * slot is available, we probe the currently pending buffers events for
+ * completion. Completed pending buffers are returned to the reserve. If so, we
+ * queue the buffer in the freed slot. If that was unsucessful (i.e., all
+ * currently pending buffers are still in use by the GPU), we have no choice but
+ * the free the buffer using the current stream.
+ *
+ * In either case, `max_reserve` represents the maxmum size of the stock. If
+ * returning a buffer would lead to the stock exeeding this quantity, the buffer
+ * is queued for destruction.
+ */
+template <class Allocator>
+class MemoryPool final {
+ public:
+  using pool_type = MemoryPool<Allocator>;
+  using alloc_type = typename Allocator::type;
+  template <class Container>
+  class Workspace {
+   public:
+    inline Workspace() : pool_{nullptr}, ctx_{nullptr} {}
+
+    inline Workspace(pool_type* pool, cudaStream_t ctx)
+        : pool_{pool}, ctx_{ctx} {}
+
+    Workspace(const Workspace&) = delete;
+    Workspace& operator=(const Workspace&) = delete;
+
+    inline Workspace(Workspace&& other)
+        : pool_{other.pool_},
+          ctx_{other.ctx_},
+          buffers_{std::move(other.buffers_)} {}
+
+    inline Workspace& operator=(Workspace&& other) {
+      if (pool_) {
+        pool_->put_raw(buffers_.begin(), buffers_.end(), ctx_);
+      }
+      pool_ = other.pool_;
+      ctx_ = other.ctx_;
+      buffers_ = std::move(other.buffers_);
+      other.pool_ = nullptr;
+      return *this;
+    }
+
+    inline ~Workspace() {
+      if (pool_) {
+        pool_->put_raw(buffers_.begin(), buffers_.end(), ctx_);
+      }
+    }
+
+    template <class T>
+    constexpr void at(const size_t n, T* ptr) const {
+      *ptr = at<T>(n);
+    }
+
+    template <class T>
+    constexpr T at(const size_t n) const {
+      return reinterpret_cast<T>(buffers_.at(n));
+    }
+
+    template <class T>
+    constexpr void get(const size_t n, T* ptr) const {
+      *ptr = get<T>(n);
+    }
+
+    template <class T>
+    constexpr T get(const size_t n) const {
+      return reinterpret_cast<T>(buffers_[n]);
+    }
+
+    constexpr alloc_type* operator[](const size_t n) const {
+      return buffers_[n];
+    }
+
+   protected:
+    pool_type* pool_;
+    cudaStream_t ctx_;
```

Comment:
Saving `stream` into the workspace can be a bad idea because the stream is not stable, it will be destroyed anytime, we should judge if it is available before each reference. I recommend getting the `stream` from the caller explicitly.

Reply:
The workspace is RAII. That is the entire concept of the resource pooling. When the function exits, the workspace is returned. If the stream is prematurely killed without synchronizing, it is a bug by the programmer. Maybe would need to check, but that that could be somewhat hard.

Reply:
Arguably, maybe there should be a check to avoid a crash, so that you do not attempt to synchronize on an already destroyed stream.

Reply:
For the said case to happen, the user would have to `std::move` the ownership of the Workspace out of the RAII context where it was created. I added a check.

The true asynchronous action is actually `collect_pending_unsafe`. We have given up on the stream already. Instead we hold an `cudaEvent_t` handle. We use `cudaEventQuery` to check it. To be honest, I don't know what happens if you `record` an `event` on a `stream` but destroy the `stream` before the `event` has occurred. But here we enter a circle. Somebody destroying a `stream`, without making sure that it has been synchronized. It is certainly possible, but I'd argue that it is clearly a programming error.